grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,83 @@
+ * The logic was only put in a service class so that it could be unit tested. Do not use it in regular flows.
+ */
+@Transactional
+class PrepaymentInvoiceMigrationService {
+
+    /**
+     * For STEP 1 of the migration.
+     *
+     * Set the amount field for all prepayment invoice items that existed before the partial invoicing
+     * feature was introduced.
+     */
+    void updateAmountForPrepaymentInvoiceItems() {
+        InvoiceType prepaymentInvoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+
+        Invoice.findAllByInvoiceType(prepaymentInvoiceType).each { Invoice prepaymentInvoice ->
+            // The amount field will be null for all pre-existing invoices since the field was not being used.
+            prepaymentInvoice.invoiceItems.find { it.amount == null }.each { InvoiceItem prepaymentInvoiceItem ->
+                prepaymentInvoiceItem.amount = computePrepaymentInvoiceItemAmount(prepaymentInvoiceItem)
+                prepaymentInvoiceItem.save(failOnError: true, flush: true)
+            }
+        }
+    }
+
+    private BigDecimal computePrepaymentInvoiceItemAmount(InvoiceItem prepaymentInvoiceItem) {
+        BigDecimal prepaymentPercent = prepaymentInvoiceItem.order.prepaymentPercent
+
+        // If the invoice item is for an adjustment.
+        OrderAdjustment orderAdjustment = prepaymentInvoiceItem.orderAdjustment
+        if (orderAdjustment) {
+            return (orderAdjustment?.canceled ? 0 : orderAdjustment.totalAdjustments) * prepaymentPercent
+        }
+
+        // Else if the invoice item is for a canceled order item.
+        OrderItem orderItem = prepaymentInvoiceItem.orderItem
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            return 0
+        }
+
+        // Else the invoice item is for a non-canceled order item.
+        return (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0) * prepaymentPercent
+    }
+
+    /**
+     * For STEP 2 of the migration.
+     *
+     * For every item in a given prepayment invoice, computes the inverse invoice item and insert it to the database
+     * by adding it to the given final invoice.
+     *
+     * This method assumes that because there is both a prepayment invoice and a final invoice, that the order
+     * associated with both invoices has been fully invoiced, and so there's no need to check each item of the
+     * prepayment invoice individually, they should ALL need to have inverse items created for them.
+     */
+    Invoice generateInverseInvoiceItems(Invoice prepaymentInvoice, Invoice finalInvoice) {
+        prepaymentInvoice.invoiceItems.each { InvoiceItem prepaymentInvoiceItem ->
+            InvoiceItem inverseItem = createInverseInvoiceItem(prepaymentInvoiceItem)
+            finalInvoice.addToInvoiceItems(inverseItem)
+        }
+        return finalInvoice.save(failOnError: true, flush: true)
```

Comment:
and this flush?

Reply:
this one is probably safe to not flush and just let GORM persist it as it would normally.